app/sseclient/client.go

```diff
@@ -0,0 +1,214 @@
+	ID    string
+	Event string
+	Data  []byte
+}
+
+type (
+	ErrorHandler func(err error, url string) error
+	EventHandler func(e *Event, url string) error
+)
+
+type Client struct {
+	URL        string
+	Retry      time.Duration
+	HTTPClient *http.Client
+	Headers    http.Header
+}
+
+var errStreamConn = errors.New("cannot connect to the stream")
+
+func New(url string) *Client {
+	return &Client{
+		URL:        url,
+		HTTPClient: &http.Client{},
+		Headers:    make(http.Header),
+	}
+}
+
+// Start connects to the SSE stream. This function will block until SSE stream is stopped.
+func (c *Client) Start(ctx context.Context, eventFn EventHandler, errorFn ErrorHandler) error {
+	timeout := c.Retry / 32
+	timer := time.NewTimer(0)
+
+	stop := func() {
+		timer.Stop()
+
+		select {
+		case <-timer.C:
+		default:
+		}
+	}
+	defer stop()
+
+	for {
+		// Function blocks here.
+		err := c.connect(ctx, eventFn)
+
+		switch {
+		case err == nil, errors.Is(err, io.EOF):
+			// Reset timeout if no error.
+			timeout = c.Retry / 32
+		case errors.Is(err, ctx.Err()):
+			// Exit function if context errored.
+			return nil
+		default:
+			// If error is not stream-related error, do not attempt retries and return the error.
+			if !errors.Is(err, errStreamConn) {
+				return errorFn(err, c.URL)
+			}
+
+			// Wait for the specified timeout (c.Retry/32 initially).
+			stop()
+			timer.Reset(timeout)
+
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				return nil
+			}
+
+			// Increase the timeout by twice for the next potential failure, up to c.Retry.
+			if timeout < c.Retry {
+				timeout *= 2
+			}
+		}
+	}
+}
+
+func (c *Client) connect(ctx context.Context, eventFn EventHandler) error {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.URL, nil)
+	if err != nil {
+		return errors.Wrap(err, "create new request")
+	}
+
+	for h, vs := range c.Headers {
+		for _, v := range vs {
+			req.Header.Add(h, v)
+		}
+	}
+	req.Header.Set("Accept", "text/event-stream")
+
+	resp, err := c.HTTPClient.Do(req)
+	if err != nil {
+		return errStreamConn
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		r := bufio.NewReader(resp.Body)
+
+		for {
+			event, err := c.parseEvent(r)
+			if err != nil {
+				return err
+			}
+
+			if len(event.Data) == 0 {
+				continue
+			}
+
+			if err := eventFn(event, c.URL); err != nil {
+				return err
+			}
+		}
+	default:
+		return errors.New("bad response status code", z.Int("status_code", resp.StatusCode))
+	}
+}
+
+func (c *Client) parseEvent(r *bufio.Reader) (*Event, error) {
+	event := &Event{
+		Event: "message",
+	}
+
+	for {
+		parts, err := formatAndValidateEvent(r)
+		if err != nil {
+			return nil, err
+		}
+		if len(parts) == 0 {
+			return event, nil
+		}
+
+		// Check response type.
+		switch string(parts[0]) {
+		case "retry":
+			ms, err := strconv.Atoi(string(parts[1]))
+			if err != nil {
+				continue
+			}
+
+			c.Retry = time.Duration(ms) * time.Millisecond
```

Comment:
shall we disconnect+reconnect after this?

Reply:
I've looked into BNs implementations and it doesn't seem they utilise this "retry" call. However, I've kept the implementation of it to be compliant with the [official docs](https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface) of event sourcing. The official docs don't really give details about what is expected from "retry", so I assume it's up to the implementer (as some other [parts of the doc](https://html.spec.whatwg.org/multipage/server-sent-events.html#concept-event-stream-reconnection-time)). Given that BNs don't utilise that, I don't think it's worth putting more effort into that...